middlewares/retry.go

```diff
@@ -32,12 +33,19 @@ func NewRetry(attempts int, next http.Handler) *Retry {
 }
 
 func (retry *Retry) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	// if we might make multiple attempts, swap the body for an ioutil.NopCloser
+	// cf https://github.com/containous/traefik/issues/1008
+	if retry.attempts > 1 {
+		body := r.Body
+		defer body.Close()
+		r.Body = ioutil.NopCloser(body)
+	}
 	attempts := 1
 	for {
 		recorder := NewRecorder()
 		recorder.responseWriter = rw
 		retry.next.ServeHTTP(recorder, r)
-		if !isNetworkError(recorder.Code) || attempts >= retry.attempts {
+		if recorder.Error == nil || attempts >= retry.attempts {
```

Comment:
>Would you consider these as network errors too?

Not all these errors are network errors. For example https://github.com/containous/oxy/blob/master/forward/fwd.go#L278.

Therefore, I think it would be better to revert to the old test `if !isNetworkError(recorder.Code) || attempts >= retry.attempts {` :)